test/test_test.jl

```diff
@@ -53,5 +53,31 @@ MPI.Waitall(reqs)
 @test isnothing(inds)
 @test isempty(stats)
 
+reqs = MPI.MultiRequest(2)
+MPI.Irecv!(recv_mesg, comm, reqs[1]; source=src, tag=src+32)
+@test !MPI.Testall(reqs)
+
+MPI.Barrier(comm)
+MPI.Isend(send_mesg, comm, reqs[2]; dest=dst, tag=rank+32, )
+
+inds = MPI.Waitsome(reqs)
+@test !isempty(inds)
+
+MPI.Waitall(reqs)
+@test MPI.Testall(reqs)
+
+reqs = MPI.NoRefMultiRequest(2)
```

Comment:
So this is where semantics get difficult. 

You will need a `GC.@preserve` from Line 70 until the Waitall. Otherwise the GC could delete recv_mesg after we exit from `Irecv!` allocate new data immediatly in the same memory and MPI will come along and scribble over the data.

Reply:
Yes, but I really want the user to be aware of what they are doing, and not get used to arguing around livetimes.  (Basically I am anticipating folks copy-and-pasting this)
E.g. a unsuspecting soul might refactor this into a `let` and then escape analysis could theoretically prove that the life time of the object ended at the `Irecv!`